server/src/main/java/org/candlepin/controller/Entitler.java

```diff
@@ -376,47 +382,67 @@ private DeveloperProducts getDeveloperPoolProducts(Owner owner, String sku) {
      *         from the adapter.
      */
     private DeveloperProducts getDevProductMap(Owner owner, String sku) {
-
         Collection<? extends ProductInfo> productsByIds = this.productAdapter
             .getProductsByIds(owner.getKey(), Arrays.asList(sku));
 
-        Collection<String> devProductIds = new ArrayList<>();
-        if (productsByIds.iterator().hasNext()) {
+        Map<String, Product> devProductMap = new HashMap<>();
+
+        if (productsByIds != null && !productsByIds.isEmpty()) {
+            // We're apparently only interested in the first product returned for the given sku
             ProductInfo devProduct = productsByIds.iterator().next();
-            devProductIds.add(devProduct.getId());
 
-            Collection<? extends ProductInfo> provided = devProduct.getProvidedProducts();
+            // Collect the dev product IDs from the potential tree of products we received
+            List<String> devProductIds = new ArrayList<>();
+            this.collectDevProductIds(devProductIds, devProduct);
 
-            if (provided != null) {
-                provided.forEach(product -> {
-                    if (product != null) {
-                        devProductIds.add(product.getId());
-                    }
-                });
-            }
-        }
+            // Do a refresh so we're all up to date here
+            log.debug("Importing products for dev pool resolution...");
 
-        log.debug("Importing products for dev pool resolution...");
-        ImportedEntityCompiler compiler = new ImportedEntityCompiler();
+            RefreshWorker refresher = this.refreshWorkerProvider.get();
+            refresher.addProducts(this.productAdapter.getProductsByIds(owner.getKey(), devProductIds));
```

Comment:
This is not directly related to your changes, but rather the previous changes in this method.
If the initial call of `productAdapter.getProductsByIds(owner.getKey(), Arrays.asList(sku))` returns the marketing product we want, I assume that when Teamnado are going to implement our current ProductInfo interface which has the new `getProvidedProducts()` method on it, then that means we already have the whole product hierarchy fetched (well, for now a single mkt product on level 0 and some eng products on level 1). Why do we need to call the adapter a second time for all the children of `sku`?

Reply:
That's a very valid point.

Today this path is still necessary to get the overall code working, since our default product service adapter hasn't been ported to the new model yet. But it would be a lot nicer to not need to do a fetch->collect->fetch step.